shared/strbuf.h

```diff
@@ -12,10 +13,24 @@ struct strbuf {
 	char *bytes;
 	size_t size;
 	size_t used;
+	bool heap;
 };
 
 void strbuf_init(struct strbuf *buf);
 
+/*
+ * Initialize strbuf with an initial buffer on stack. The @sz must be a build-time
+ * constant, as if the buffer had been declared on stack.
+ */
+#define strbuf_init_with_stack(buf__, sz__)            \
+	({                                             \
+		assert_cc(__builtin_constant_p(sz__)); \
+		(buf__)->bytes = alloca(sz__);         \
```

Comment:
Was thinking about adding `-Walloca` because of the obvious fallouts with it. The `__builtin_constant_p` does help with those, but I wonder if platforms like Android/macOS or other C runtimes don't impose restrictions on its use.

As a potential replacement we could use a `static char foo[sz__];`, which will invalidate the documented caveat - aka "It may also not survive a return from current function if it was initialized with stack space" can go.

This will grow the BSS (non-issue IMHO) although I wonder if it'll have other side-effects.

Mostly thinking out loud, than an objection.

Reply:
The static variable is a no go. We don't ever want a a "stack variable" to implicitly survive a return from the function.

Alternative IMO is what scratchbuf does with the additional variable. We could define a DECLARE_STRBUF_WITH_STACK that does that and hide the variable under a unique name.

Reply:
> Was thinking about adding `-Walloca` because of the obvious fallouts with it.

in this case we'd just add some _Pragma() to push/pop -Walloca. There are uses for it, this being one of them.  What I like about not using alloca() here is actually that the buffer would be defined together with the declaration rather than together with strbuf_init_with_stack().

Reply:
Same - scope and lifetime of using alloca() is a bit tricky. The DECLARE_STRBUF_WITH_STACK approach seems reasonable IMHO.

Reply:
> This will grow the BSS (non-issue IMHO) although I wonder if it'll have other side-effects.

A side-effect occurs in multi-threaded environments, because then multiple threads share the same static variable. Definitely undefined behavior, so that wouldn't be an option for library code.

Reply:
The good thing about `alloca` hidden in a macro here is that we have an "anonymous" stack memory which we can never accidentally access again like it happened before in depmod. I think it's the first reasonable example of `alloca` usage I've seen so far and with these checks in place, it cannot eventually overflow.

I'm more concerned about the lifetime here. And it's no difference if it's a named and easily accessible array on the stack or something hidden behind a macro.

Reply:
I will add a commit on top with one of the alternatives I have in mind, then we decide and I can squash or leave it as is.

Reply:
The buffer variable will still be hidden from normal use.

Reply:
I like the alloca-free approach as well. Actually a bit more because it's without alloca. :)